mpqp/qasm/qasm_to_mpqp.py

```diff
@@ -0,0 +1,349 @@
+            if tokens[idx].type == 'ID' and tokens[idx + 1].type == 'SEMICOLON':
+                c_targets = None
+                idx += 1
+                break
+            if check_Id(tokens, idx):
+                raise SyntaxError(
+                    ' '.join(str(token.value) for token in tokens[idx : idx + 4])
+                )
+            c_targets.append(tokens[idx + 2].value)
+            idx += 4
+            if tokens[idx].type == 'COMMA':
+                idx += 1
+
+    circuit.add(BasisMeasure(targets, c_targets))
+    return idx + 1
+
+
+def _TokenBarrier(circuit: QCircuit, tokens: list[LexToken], idx: int) -> int:
+    idx += 2
+    if tokens[idx].type != 'SEMICOLON':
+        raise SyntaxError(f"Barrier: {idx} {tokens[idx]}")
+    circuit.add(Barrier())
+    return idx + 1
+
+
+def _TokenGate(circuit: QCircuit, tokens: list[LexToken], idx: int) -> int:
+    token = tokens[idx]
+    idx += 1
+    token_value = token.value.lower()
+    if token_value in single_qubits_gate_qasm:
+        return _Gate_single_qubits(circuit, token_value, tokens, idx)
+    elif token_value in two_qubits_gate_qasm:
+        return _Gate_two_qubits(circuit, token_value, tokens, idx)
+    elif token_value in one_parametrized_gate_qasm:
+        return _Gate_one_parametrized(circuit, token_value, tokens, idx)
+    elif token_value in u_gate_qasm:
+        return _Gate_U(circuit, token_value, tokens, idx)
+    elif token_value in two_qubits_parametrized_gate_qasm:
+        return _Gate_two_qubits_parametrized(circuit, token_value, tokens, idx)
+    elif token_value == "ccx":
+        return _Gate_tof(circuit, tokens, idx)
+    else:
+        raise SyntaxError(f"TokenGate: {idx} {token.value}")
+
+
+def _Gate_single_qubits(
+    circuit: QCircuit, gate_str: str, tokens: list[LexToken], idx: int
+) -> int:
+    if tokens[idx].type == 'ID' and tokens[idx + 1].type == 'SEMICOLON':
+        for i in range(circuit.nb_qubits):
+            circuit.add(single_qubits_gate_qasm[gate_str](i))
+        return idx + 2
+
+    multi = False
+    while tokens[idx].type != 'SEMICOLON':
+        if multi:
+            if tokens[idx].type != 'COMMA':
+                raise SyntaxError(f"Gate_single_qubits: {idx} {tokens[idx]}")
+            idx += 1
+        if check_Id(tokens, idx):
+            raise SyntaxError(
+                f'Gate_single_qubits: {" ".join(str(token.value) for token in tokens[idx : idx + 4])}'
+            )
+        circuit.add(single_qubits_gate_qasm[gate_str](tokens[idx + 2].value))
+        idx += 4
+        multi = True
+    return idx + 1
+
+
+def _Gate_two_qubits_parametrized(
+    circuit: QCircuit, gate_str: str, tokens: list[LexToken], idx: int
+) -> int:
+    if tokens[idx].type != 'LPAREN':
+        raise SyntaxError(f"Gate_one_parametrized: {idx} {tokens[idx]}")
+    idx += 1
+    parameter, idx = _eval_expr(tokens, idx)
+    if gate_str == 'cp':
+        parameter = np.log2(np.pi / parameter) + 1
```

Comment:
Looks like you handle here the parameter translation between the phase of `cp` and the order `k` of `CRk`.
However here your `k` will probably be a float, while `CRk` takes an integer I think.
As we discussed for the moment we can maybe round and put a todo for the future.

Reply:
yes, CRk can take float